todolist

myaak opened the pull request:
mobx + styled components потрогал

myaak committed:
todolist mobx/styledcomponents

myaak committed:
quickfix

Mary-fox commented:
Есть ошибка при реализации(я сама с ней столкнулась). Проверь работоспособность кнопки edit в отфильтрованном списке... У тебя редактируется первый не отфильтрованный пункт,а не отфильтрованный.  Пока протестировала только саму тудушку, код не смотрела

myaak committed:
fix: поправил неверную отрисовку при эдите

myaak committed:
remanaged components

Mary-fox commented on src/components/TodoItem/TodoItem.tsx:
```diff
@@ -0,0 +1,58 @@
+const TodoItem: React.FC<ITodoItem> = observer(({todo, onCheck}) => {
+  const {title, completed} = todo;
+
+  const [editing, setEditing] = useState<boolean>(false);
```
Хорошей практикой наверное будет типизировать и  деструктурированные свойства объекта

Mary-fox commented on src/components/TodoAddForm/TodoAddForm.tsx:
```diff
@@ -0,0 +1,27 @@
+    onAdd(newTitle);
+    setNewTitle('');
+  }
+
```
можно еще прописать условие,чтобы нельзя было добавлять пустые задачи.. Если поставить пару пробелов в инпуте, то добавляется пустая задача в список

Mary-fox commented on src/components/TodoList/TodoListComponent.tsx:
```diff
@@ -0,0 +1,25 @@
+        />
+      ))  
+      }
+    </ul>
```
я не очень понимаю,почему везде id,а тут используется индекс

G00syara commented:
Тестировал todo, советовал бы поставить в input, который отвечает за добавление новой todo ограничение, на количество символов, либо сделать перенос текста на следующую "строку", после определённого расстояния, прикрепляю ссылку на скрин
https://imgur.com/U6Mz7Us

G00syara commented on src/components/TodoItem/TodoItem.tsx:
```diff
@@ -0,0 +1,58 @@
+  const [editing, setEditing] = useState<boolean>(false);
+  const [newTitle, setNewTitle] = useState<string>('');
+
+  const handleSaveResult = () => {
+    TodoList.renewTitle({...todo, title: newTitle});
+    setEditing(false);
+  }
```
Попробуй у кнопки Save сделать проверку на введённые символы, потому что можно отредактировать todo так, чтобы не осталось символов, прикрепляю ссылку на фото
https://imgur.com/CJ3tCvj

G00syara commented:
У тебя происходить рендер всех todo, на каждое действие, например:
Ты добавляешь новую todo, через addTodo, у тебя рендерится и первая todo и та которую ты добавил
Ты удаляешь todo, у тебя ре-рендерятся все твои другие todo
Можешь проверить это добавив в TodoItem  console.log(`render item ${todo.id} ${title}`)
Прикрепляю скрины
https://imgur.com/GCiQKzz
https://imgur.com/HX2vyF1
У меня была такая же проблема, я её исправил нижеприведённым образом:
- Обернул отфильтрованные todo в useMemo() (const handleFilterTodos = useMemo(()=>,[ ])
- Обернул функции, которые передаются в TodoItem, в useCallback() (const handleAddTodo = useCallback(()=>,[ ]);

> P.S. Не уверен, что useMemo() и useCallback() тебе обязательно нужно будет использовать, т.к. ты используешь mobx, который по идее имеет функционал для упрощённого решения этой проблемы.
> P.S.S Если я правильно понял, то вместо useMemo() можно использовать observer, а вместо useCallback() можно использовать action

- И у экспортируемых компонентов добавлял React.memo() (export default React.memo(TodoItem))

Вроде ничего не забыл, по этому поводу

G00syara commented on src/assets/react.svg:
Не нашёл применения этой картинки в коде

G00syara commented on src/components/TodoAddForm/TodoAddForm.tsx:
```diff
@@ -0,0 +1,27 @@
+import { Button } from "../Buttons/Buttons.style";
+import { AddForm, AddFormInput } from "./TodoAddForm.style";
+
+interface ITodoAddForm {
+  onAdd: (title: string) => void;
+}
```
Предлагаю типизировать onAdd так: (title: Todo['title']) => void
Предварительно импортировав ITodo в код (import Todo from "../../store/Todo";)

G00syara commented on src/components/TodoItem/TodoItem.tsx:
```diff
@@ -0,0 +1,58 @@
+}
+
+const TodoItem: React.FC<ITodoItem> = observer(({todo, onCheck}) => {
+  const {title, completed} = todo;
```
Предлагаю типизировать через Omit<Todo, ['id']>

myaak replied:
норм идея

G00syara commented on src/components/TodoItem/TodoItem.tsx:
```diff
@@ -0,0 +1,58 @@
+  const handleCancelEditing = useCallback(() => {
+    setEditing(false);
+    setNewTitle(title);
+  }, [])
```
Не понимаю зачем нужна здесь строка setNewTitle(title), меняя её содержание и удаляя её, ничего не менялось

G00syara commented on src/components/TodoItem/TodoItem.tsx:
Во время фильтрации, todo у которых нажата кнопка edit, не корректно отображаются, пример:
Представим, что todo состоит из 3 элементов:

- JavaScript [Edit]
- TypeScript [Edit]
- Python [Edit]

Если мы нажмём у JavaScript, кнопку [Edit] и начнём вписывать в фильтр слово Type, то окно редактирования перенесётся на TypeScript, а у JavaScript пропадёт

G00syara commented on src/components/TodoList/TodoListComponent.tsx:
```diff
@@ -0,0 +1,25 @@
+const TodoListComponent: React.FC<ITodoListComponent> = observer(({todos}) => {
+  return (
+    <ul>
+      {todos.map((todo: Todo, index: number) => (
+        <TodoItem
+          key={index}
+          todo={todo}
```
Предложил бы вместо key={index} написать key={todo.id}

myaak replied:
пофиксил

G00syara commented on src/components/TodoList/TodoListWrapper.tsx:
```diff
@@ -0,0 +1,46 @@
+      return;
+    }
+
+    const filteredArray: Todo[] = 
+      TodoList.todos.filter((item: Todo) => item.title.toLowerCase().includes(filterString.toLowerCase()));
+
+    setTodos(filteredArray);
```
Предложил бы назвать не filteredArray, а filteredTodos, тут же происходить фильтрация тудушек, а не какого то случайного массива

myaak replied:
пофиксил

myaak committed:
fix: пофиксил пустой туду после сейва и добавил ограничение

myaak committed:
fix: переименовал переменные, добавил тип объекту

myaak committed:
fix: еще раз поправил длину названия

myaak committed:
fix: не чекался чекбокс

Mary-fox commented on src/components/TodoList/TodoListWrapper.tsx:
```diff
@@ -0,0 +1,51 @@
+    TodoList.addTodo(title);
+    handleFilterTodos();
+  })
+
+  useEffect(() => {
+    handleFilterTodos();
+  }, [filterString, TodoList.todos]);
+
+  return (
```
Мне кажется вместо useEffect можно еще reaction из Mobx использовать.

stark355 commented on src/main.tsx:
поломался линтинг, в частиности табы и ;

stark355 commented on src/components/TodoList/TodoListWrapper.tsx:
```diff
@@ -0,0 +1,51 @@
+import { useCallback, useEffect, useState } from 'react';
+import Todo from "../../store/Todo";
+import TodoListComponent from './TodoListComponent';
```
где-то одинарные кавычки, где-то двойные. прогони проект через линтер/преттиер, больше пока не буду обращать внимание на косяки с оформлением

stark355 commented on src/store/TodoList.ts:
```diff
@@ -0,0 +1,46 @@
+    this.todos[index].completed = !completed;
+  }
+
+  renewTitle({id, title}: Todo) {
```
имхо так себе название

stark355 commented on src/store/TodoList.ts:
```diff
@@ -0,0 +1,46 @@
+    this.todos = this.todos.filter((item: Todo) => item.id !== id);
+  }
+
+  completeTodo({id, completed}: Todo) {
```
неподходящее название. completeTodo говорит о том, что ты завершаешь задание, хотя в коде ты можешь наоборот снять галку

stark355 replied:
я увидел, ты в коде прокидвыаешь сам туду. тебе нет смысла искать его заново. просто модифицируй этот туду
todo.checked = !todo.checked

stark355 commented on src/models/ITodo.ts:
вместо явного описания ITodo можно использовать такое https://www.typescriptlang.org/docs/handbook/2/classes.html#constructor-signatures
из плюсов - если поменяешь класс, то поменяется и тип

stark355 commented on src/store/TodoList.ts:
```diff
@@ -0,0 +1,46 @@
+  }
+
+  completeTodo({id, completed}: Todo) {
+    const index = this.todos.findIndex((item: Todo) => item.id === id);
```
стоит проверять, что индекс > -1

stark355 commented on src/store/TodoList.ts:
```diff
@@ -0,0 +1,46 @@
+  }
+
+  renewTitle({id, title}: Todo) {
+    this.todos = this.todos.map((item: Todo) => item.id === id ? {...item, title: title} : item);
```
ты можешь как и в методе выше найти элемент и его изменить

UPD: тебе не нужно его искать или что-либо делать с массивом. у тебя уже есть прокинутый в параметр элемент, просто модифицируй его

stark355 commented on src/components/TodoList/TodoListWrapper.tsx:
```diff
@@ -0,0 +1,51 @@
+  })
+
+  useEffect(() => {
+    handleFilterTodos();
```
логичнее было бы сделать мемоизированное значение filteredTodos, чем императивно вызывать фильтрацию

stark355 commented on src/components/TodoList/TodoListWrapper.tsx:
```diff
@@ -0,0 +1,51 @@
+  const handleAddTodo = action((title: string) => {
+    if (title === '') return;
+    TodoList.addTodo(title);
+    handleFilterTodos();
```
в целом не очень верный подход, об этом написал ниже. но если использовать его, то все равно непонятно зачем тут handleFilterTodos(). у тебя ниже есть useEffect, который в любом случае вызовет фильтрацию

stark355 commented on src/components/TodoList/TodoListComponent.tsx:
```diff
@@ -0,0 +1,34 @@
+  todos: Todo[];
+}
+
+const TodoListComponent: React.FC<ITodoListComponent> = observer(({todos}) => {
```
здесь ты обсервер вешаешь при объявлении функции, в TodoAddForm - оборачиваешь при экспорте. стоит придерживаться одного стиля. в целом неважно какого
P.S. лично мне нравится обсервер при экспорте навешивать

stark355 commented on src/components/TodoList/TodoListComponent.tsx:
```diff
@@ -0,0 +1,34 @@
+import { useEffect } from "react";
+import { action } from "mobx";
+
+interface ITodoListComponent {
```
такой вариант тоже годный, но чаще встречаю вариант TodoListComponentProps

stark355 commented on src/components/TodoList/TodoListComponent.tsx:
```diff
@@ -0,0 +1,34 @@
+import Todo from "../../store/Todo";
+import TodoItem from "../TodoItem/TodoItem";
+import { observer } from "mobx-react";
```
ругается, что не может найти модули, не ищет в node_modules
в tsconfig стоит прописать moduleResolution: "node"

stark355 commented on src/components/TodoList/TodoListComponent.tsx:
```diff
@@ -0,0 +1,34 @@
+    TodoList.completeTodo(todo);
+  });
+
+  useEffect(() => {
```
это зачем?

stark355 commented on src/components/TodoList/TodoListComponent.tsx:
```diff
@@ -0,0 +1,34 @@
+        <TodoItem
+          key={todo.id}
+          todo={todo}
+          onCheck={action(() => handleCompleteTodo(todo))}
```
не оч понял зачем тут тоже оборачивать в action

stark355 commented on src/components/TodoItem/TodoItem.tsx:
```diff
@@ -0,0 +1,68 @@
+
+interface ITodoItem {
+  todo: Todo;
+  onCheck: () => void;
```
в общем-то ты можешь эти методы не прокидывать как пропы, а забирать из стора

stark355 commented on src/components/TodoItem/TodoItem.tsx:
```diff
@@ -0,0 +1,68 @@
+}
+
+const TodoItem: React.FC<ITodoItem> = observer(({todo, onCheck}) => {
+  const {title, completed}: Omit<Todo, "id"> = todo;
```
ты делаешь деструктуризацию объекта, пофиг какого. при деструктуризации ты заберешь только нужные поля. зачем тут Omit?

stark355 commented on src/components/TodoItem/TodoItem.tsx:
```diff
@@ -0,0 +1,68 @@
+
+  const handleSaveResult = action(() => {
+    if (newTitle === '') return;
+    TodoList.renewTitle({...todo, title: newTitle});
```
опять же, просто прокидывай todo и title, тебе не нужен тут spread

stark355 commented on src/components/TodoItem/TodoItem.tsx:
```diff
@@ -0,0 +1,68 @@
+  return (
+    <TodoWrapper>
+      { editing ? 
+        <input type="text" 
```
можно завернуть в form, чтобы по enter сохранялось

stark355 commented on src/components/TodoItem/TodoItem.tsx:
```diff
@@ -0,0 +1,68 @@
+    setNewTitle(title);
+  });
+
+  useEffect(() => {
```
объясни плс назначение этого useEffect, когда предполагаешь что он выполнится?

stark355 commented on src/components/TodoItem/TodoItem.tsx:
```diff
@@ -0,0 +1,68 @@
+        :
+        <>
+          <Checkbox checked={completed} onChange={onCheck}></Checkbox>
+          <TodoTitle done={String(completed)} onClick={() => setEditing(true)}>
```
если делаешь элемент кликабельным - поставь ему pointer на hover

stark355 commented on src/components/TodoItem/Buttons.tsx:
```diff
@@ -0,0 +1,28 @@
+
+interface ITodoItemButtons {
+  editing: boolean;
+  saveResult: () => void;
```
вещи, касающиеся стора можно вытащить из стора, сюда можно передать только например id элемента или (лучше) todo

myaak replied:
ну я тут просто сделал онклик ивенты на кнопки TodoItem'a и все функции вынес уже в компонент повыше чтобы все было в одном месте

myaak committed:
refactor: правки по ревью

Изменил функции и их названия в классе TodoList, поправил observer(вынес его в export компонента), добавил форму на изменение title, прошелся приттиером и линтером, удалил ненужные useEffect

myaak committed:
refactor: изменил типы в todoitem

myaak committed:
refactor: добавил pointer на hover (title)

myaak committed:
fix: исправил filter убрал функцию фильтра и использовал useMemo для массива filteredTodo

myaak committed:
fix: типизировал новое туду

myaak committed:
fix: избавился от state TODOS и использую только filteredTodos для отрисовки

stark355 commented on src/components/TodoList/TodoListWrapper.tsx:
```diff
@@ -9,29 +9,23 @@ import { action } from "mobx";
 
   useEffect(() => {
-    handleFilterTodos();
+    setTodos(filteredTodos);
```
ты можешь дальше передавать в дочерние компоненты сразу filteredTodos, тебе по идее этот стейт не нужен

stark355 replied:
upd: увидел, что в след. коммите ты это поправил. неактуально

stark355 commented on src/store/TodoList.ts:
```diff
@@ -25,7 +25,7 @@ class TodoList {
       completed: false
     };
-    this.todos.push(new Todo(newTodo));
+    this.todos = [...this.todos, newTodo];
```
в принципе пуш должен был тоже работать, так у тебя не стали бы заменяться лишние ссылки и было бы меньше ререндеров
